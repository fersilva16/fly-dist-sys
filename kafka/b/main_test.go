package main

import (
	"gossip-gloomers/testutils"
	"testing"

	"github.com/gkampitakis/go-snaps/snaps"
	maelstrom "github.com/jepsen-io/maelstrom/demo/go"
	"github.com/stretchr/testify/require"
)

func Test(t *testing.T) {
	require := require.New(t)

	node = maelstrom.NewNode()

	link := testutils.NewLink(node)
	client := testutils.NewClient("c0", link)
	kv := testutils.NewLinKV(link)

	go main()

	err := client.InitNode("n0", []string{"n0"})

	require.NoError(err)

	sendOutput, err := client.RPC(SendRequest{
		MessageBody: maelstrom.MessageBody{
			Type:  "send",
			MsgID: 2,
		},

		Key: "0",
		Msg: 83,
	})

	require.NoError(err)

	snaps.MatchSnapshot(t, sendOutput)

	pollOutput, err := client.RPC(PollRequest{
		MessageBody: maelstrom.MessageBody{
			Type:  "poll",
			MsgID: 3,
		},

		Offsets: map[string]int{
			"0": 0,
		},
	})

	require.NoError(err)

	snaps.MatchSnapshot(t, pollOutput)

	sendOutput2, err := client.RPC(SendRequest{
		MessageBody: maelstrom.MessageBody{
			Type:  "send",
			MsgID: 4,
		},

		Key: "0",
		Msg: 84,
	})

	require.NoError(err)

	snaps.MatchSnapshot(t, sendOutput2)

	pollOutput2, err := client.RPC(PollRequest{
		MessageBody: maelstrom.MessageBody{
			Type:  "poll",
			MsgID: 5,
		},

		Offsets: map[string]int{
			"0": 1,
		},
	})

	require.NoError(err)

	snaps.MatchSnapshot(t, pollOutput2)

	err = client.Write(CommitOffsetsRequest{
		MessageBody: maelstrom.MessageBody{
			Type:  "commit_offsets",
			MsgID: 6,
		},

		Offsets: map[string]int{
			"0": 1,
		},
	})

	require.NoError(err)

	err = kv.HandleWrite("commit-0", float64(1))

	require.NoError(err)

	commitOffsetsOutput, err := client.Read()

	require.NoError(err)

	snaps.MatchSnapshot(t, commitOffsetsOutput)

	err = client.Write(ListCommittedOffsetsRequest{
		MessageBody: maelstrom.MessageBody{
			Type:  "list_committed_offsets",
			MsgID: 7,
		},

		Keys: []string{
			"0",
		},
	})

	require.NoError(err)

	err = kv.HandleRead("commit-0", float64(1))

	require.NoError(err)

	listCommittedOffsetOutput, err := client.Read()

	require.NoError(err)

	snaps.MatchSnapshot(t, listCommittedOffsetOutput)
}
