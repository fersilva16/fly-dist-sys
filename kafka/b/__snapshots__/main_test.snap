
[Test - 1]
{"src":"n0","dest":"c0","body":{"in_reply_to":2,"offset":0,"type":"send_ok"}}
---

[Test - 2]
{"src":"n0","dest":"c0","body":{"in_reply_to":3,"msgs":{"0":[[0,83]]},"type":"poll_ok"}}
---

[Test - 3]
{"src":"n0","dest":"c0","body":{"in_reply_to":4,"offset":1,"type":"send_ok"}}
---

[Test - 4]
{"src":"n0","dest":"c0","body":{"in_reply_to":5,"msgs":{"0":[[1,84]]},"type":"poll_ok"}}
---

[Test - 5]
{"src":"n0","dest":"c0","body":{"in_reply_to":6,"type":"commit_offsets_ok"}}
---

[Test - 6]
{"src":"n0","dest":"c0","body":{"in_reply_to":7,"offsets":{"0":1},"type":"list_committed_offsets_ok"}}
---
